gamercade_console/src/console/contexts/draw_context.rs

```diff
@@ -261,6 +231,36 @@ struct XCord(usize);
 #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
 struct YCord(usize);
 
+trait ScreenCoordinate<T> {
+    fn on_screen(&self, screen: &T) -> bool;
+}
+
+impl<T> ScreenCoordinate<T> for XCord
+where
+    T: Deref<Target = DrawContext>,
+{
+    fn on_screen(&self, screen: &T) -> bool {
+        let width: usize = match screen.width().try_into() {
+            Ok(w) => w,
+            _ => return false,
+        };
+        self.0 < width
+    }
+}
+
+impl<T> ScreenCoordinate<T> for YCord
+where
+    T: Deref<Target = DrawContext>,
+{
+    fn on_screen(&self, screen: &T) -> bool {
+        let height: usize = match screen.height().try_into() {
+            Ok(h) => h,
+            _ => return false,
+        };
+        self.0 < height
+    }
+}
+
```

Comment:
This is an interesting abstraction! I like the idea, but I wonder if there might be something slightly better.

I think it's worth it exploring something like a `YCord` function `try_for_screen(screen: Screen) -> Option<Self>;` type of function. So that we can do easy conversions from raw `i32` values from wasm, and convert them into valid-at-compile-time X or Y coordinates. IE, if we have a YCord or XCord, we know it's safe.

This could then reduce some of the extra checks when trying to validate pixel coordinates to pixel indices and make the code easier to read.

What do you think?

Reply:
Seems a good idea!
I'll work on this :+1: